Ixxat CAN FD support

fjburgos opened the pull request:
Add experimental support for CAN FD in IXXAT device(s).

Tryed and worked with "Ixxat USB-to-CAN FD".

fjburgos commented:
Please, tell me if any action is required from my side.

zariiii9003 commented:
Hello @fjburgos, thank you for your contribution.
Could you use the `fd` parameter of the normal ixxat interface to keep the API consistent? Defining a new separate inteface seems unnecessary.

fjburgos commented:
> 
> 
> Hello @fjburgos, thank you for your contribution.
> Could you use the `fd` parameter of the normal ixxat interface to keep the API consistent? Defining a new separate inteface seems unnecessary.

I agree, the problem was I have no way to check the previous interface still works, but I think I can try to do it with minimal changes to the original file (can/interfaces/ixxat/canlib.py).

fjburgos commented:
I was surprised to see that similar pull request was done 3 days ago (https://github.com/hardbyte/python-can/pull/1126).
I think that the main difference at this point is that in this pull request, the original implementation (based on vcinpl.dll) is being used when `fd` option is False, which has advantages and disadvantages.
One advantage is that backward-compatibility could be better.
One disadvantage is that some duplicated code exists (due to my approach). A refactor could be possible, but I would not do it without being able to test with a proper device that nothing gets broken.
One thing I don't know is if vcinpl2.dll can handle all IXXAT devices or only those that support CAN-FD. That's why I wanted to leave the original implementation available.
Feel free to choose what you think is better for the project.
@ausserlesh, you are welcome to share you opinion, I think it will be valuable.

ausserlesh commented:
Thanks for your thoughts, @fjburgos, I actually implemented this quite a while ago, and only saw your PR after creating mine. Yes, fully agree with the pros and cons you mentioned. Regarding vcinpl2 it indeed seems that we might consider loading vcinpl instead of vcinpl2 if fd is False (see https://opensource.lely.com/canopen/docs/cross-compilation/), which then leads again to a similar solution like the one in your PR, maybe with a different level of abstraction (hide implementation behind fd flag rather than exposing it as a separate device). Unfortunately I only have a CAN-FD device available for testing.

fjburgos commented:
I did the change in the last commit to use the fd option from interface (single ixxat interface, fd parameter decides which dll is loaded).
Thanks for your comments, this is my first conttribution to an open source project and one of the objectives was to share points of view.

ausserlesh commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,920 @@
+        (HANDLE, structures.PCANLINESTATUS2),
+        __check_status,
+    )
+    # EXTERN_C HRESULT VCIAPI canControlGetCaps( IN  HANDLE           hCanCtl, OUT PCANCAPABILITIES pCanCaps );
```
Maybe consider updating the interface descriptions inline documentation for the changed functions.

fjburgos replied:
Done

ausserlesh commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,920 @@
+        # frequency [1/s] = dwClockFreq / dwTscDivisor
+        # We explicitly cast to float for Python 2.x users
+        self._tick_resolution = float(
+            self._channel_capabilities.dwTscClkFreq  # TODO confirm
```
Think this looks OK, maybe remove the TODO.

fjburgos replied:
Removed

ausserlesh commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,920 @@
+            arbitration_id=self._message.dwMsgId,
+            dlc=data_len,
+            data=self._message.abData[:data_len],
+            channel=self.channel,
```
Maybe also initialize `is_fd`, `bitrate_switch`, `error_state_indicator`, at least that's the ones I found (maybe not complete). What do you think?

fjburgos replied:
Good point! Changed.

ausserlesh commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,920 @@
+        # This system is not designed to be very efficient
+        message = structures.CANMSG2()
+        message.uMsgInfo.Bits.type = constants.CAN_MSGTYPE_DATA
+        message.uMsgInfo.Bits.rtr = 1 if msg.is_remote_frame else 0
```
Maybe initialize `esi` - not sure though.

fjburgos replied:
I agree, also changed.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,920 @@
+            1,
+            txFifoSize,
+            1,
+            0,
```
I'm not aware about message filtering support in python-can, are you talking about implementing some interface (common for all interfaces), so that we enable hardware optimizations for set_filters function?
I could imitate the initialization from yout pull erquest, but I'm not sure about what is better. If you have a strong opinion, I can change it.
We can also leave it for another pull request in next iterations.

ausserlesh replied:
I agree with your proposal to leave it for another iteration, as I don't really have a strong opinion on that one. Thank you.

zariiii9003 reviewed (changes requested):
Thank you for your effort. This is a nice addition.

This is quite difficult to review since it is a lot of code and i don't have the interface to test it. Generally i would like more type hints and pythonic variable names.

Also i'm not sure whether we can copy the c function signatures from proprietary code. You have quite a lot of comments with the original c code.

But other than the `can_filters` issue this is acceptable.

fjburgos commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -422,427 +28,36 @@ def __init__(self, channel, can_filters=None, **kwargs):
             UniqueHardwareId to connect (optional, will use the first found if not supplied)
 
+        : param bool fd:
+        Default False, enables CAN-FD usage.
```
Changed

fjburgos commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -422,427 +28,36 @@ def __init__(self, channel, can_filters=None, **kwargs):
-            return 0
+        """
+        if kwargs.get("fd", False):
+            self.bus = vcinpl2.IXXATBus(channel, can_filters=None, **kwargs)
```
Ups, thank you, changed.

fjburgos commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -422,427 +28,36 @@ def __init__(self, channel, can_filters=None, **kwargs):
+            self.bus = vcinpl2.IXXATBus(channel, can_filters=None, **kwargs)
         else:
-            return 1
+            self.bus = vcinpl.IXXATBus(channel, can_filters=None, **kwargs)
```
Also changed.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+
+log = logging.getLogger("can.ixxat")
+
+from time import perf_counter as _timer_function
```
Done.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+    log.warning("IXXAT VCI library does not work on %s platform", sys.platform)
+
+
+def __vciFormatErrorExtended(library_instance, function, HRESULT, arguments):
```
Changed.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+    )
+
+
+def __vciFormatError(library_instance, function, HRESULT):
```
Changed both.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+    """
+    if isinstance(result, int):
+        # Real return value is an unsigned long
+        result = ctypes.c_ulong(result).value
```
Removing the line causes a bug, the purpose is to get an unsigned value (sometimes result is negative, and needs to be reinterpreted as unsigned).
Maybe there are other ways to do the same, for example, bitmasking with 0xFFFFFFFF (maybe less portable), but I did not measure the times, so, I think it's safer to leave it as it is.

zariiii9003 replied:
I see, but that just means that the function prototypes are declared incorrectly. See here for example:
```python3
    # HRESULT VCIAPI vciEnumDeviceOpen( OUT PHANDLE hEnum );
    _canlib.map_symbol("vciEnumDeviceOpen", ctypes.c_long, (PHANDLE,), __check_status)
```

The return value must be `HRESULT`, but the prototype is declared with `c_long`.

fjburgos replied:
I realized that there is also a ctypes.HRESULT type, but using it was causing errors (probably more adaptions required, I didn't dedicate too much time to investigate).

Using `c_ulong` instead of c_long allows to remove the check.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+            )
+        log.info("CAN Filters: %s", can_filters)
+        log.info("Got configuration of: %s", kwargs)
+        # Configuration options
```
Changed only keyword arguments, but I left the parameter names unchanged (no snake_case). I'm worried in case changing them breaks users code, but if you are convinced, I don't mind to change it after you feedback.

fjburgos replied:
Interesting, thanks, changed.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+        # the message in ticks. The resolution of a tick can be calculated from the fields
+        # dwClockFreq and dwTscDivisor of the structure  CANCAPABILITIES in accordance with the following formula:
+        # frequency [1/s] = dwClockFreq / dwTscDivisor
+        # We explicitly cast to float for Python 2.x users
```
Removed.

fjburgos commented on can/interfaces/ixxat/canlib_vcinpl2.py:
```diff
@@ -0,0 +1,948 @@
+        _canlib.vciDeviceClose(self._device_handle)
+
+
+class CyclicSendTask(LimitedDurationCyclicSendTaskABC, RestartableCyclicTaskABC):
```
I have no idea ^^'

zariiii9003 replied:
i guess we will know if someone complains

fjburgos commented:
Hi @zariiii9003,

I tried to change most of the requested changes.

Please, note that a big part of the code comes from previous vcinpl implementation, my work was more focused in adapting the calls to the driver API, this is why I cannot answer your question about CyclicSendTask.

I checked that it still works after the changes.

Thank you very much for your review, I'll wait for your feedback.

zariiii9003 commented on can/interfaces/ixxat/canlib_vcinpl.py:
```diff
@@ -56,45 +56,47 @@
 
-def __vciFormatErrorExtended(library_instance, function, HRESULT, arguments):
+def __vciFormatErrorExtended(
+    library_instance: CLibrary, function: Callable, vret: int, args: tuple
```
`tuple` only works in python 3.9 and newer. You`could use `typing.Tuple`

fjburgos replied:
changed

fjburgos commented:
Hi @zariiii9003 

I realized I did not read two comments you did, so, I included some of your suggestions in the last commit.

I've also added some optional parameters (TSEG1, TSEG2 and SJW) for can-fd.

zariiii9003 commented on doc/interfaces/ixxat.rst:
```diff
@@ -80,3 +91,4 @@ explicitly instantiated by the caller.
 
 RX and TX FIFO sizes are configurable with ``rxFifoSize`` and ``txFifoSize``
```
```suggestion
RX and TX FIFO sizes are configurable with ``rx_fifo_size`` and ``tx_fifo_size``
```

zariiii9003 commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -1,414 +1,20 @@
-    }
-
-    def __init__(self, channel, can_filters=None, **kwargs):
+    def __init__(self, channel, can_filters=None, fd=False, **kwargs):
```
Can you improve the signature and the docstring of this `__init__()` method? This one will be shown in the docs. Otherwise this looks fine :)

fjburgos replied:
Done.

fjburgos commented:
Thank you for your comments @zariiii9003 :)

zariiii9003 commented:
@felixdivo can you do the second review?

marcel-kanter commented:
@fjburgos Nice work. But the extended argument is not used in the init of vcimpl.py
Please set the default to True, because the old implementation opens the interface with

constants.CAN_OPMODE_EXTENDED

so extended ids are allowed by default.

felixdivo commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -1,414 +1,40 @@
+        channel: int,
+        can_filters=None,
+        receive_own_messages: bool = False,
+        unique_hardware_id: int = None,
```
```suggestion
        unique_hardware_id: Optional[int] = None,
```

fjburgos replied:
Changed.

felixdivo commented on test/test_interface_ixxat_fd.py:
```diff
@@ -0,0 +1,62 @@
+"""
+Unittest for ixxat interface using fd option.
```
Tests are always welcome! If you want, you can also move the hardware test to `test/back2back_test.py` and let it derive from `Back2BackTestCase`. That way, a suite of generic tests are run on the device and you don't have to write any more for this interface specifically, but still get the benefits of testing. The only change would probably be to rename the `setUp()` method (see [here](https://stackoverflow.com/q/52155375/3753684)).

fjburgos replied:
I had problems with the testing I found previously (test_interface_ixxat.py).
In the begining, I had problems to run the tests. And after being able to run them, none passed.
My experience in testing in python is small (I'd like to have the oportunity to improve that), but I don't know if tests were broken, something is wrong on my environment, or I'm doing something wrong... In any case, existing tests were not testing important things (unless I'm missing something, I only saw tests checking that exceptions were triggered if some functions were called in invalid states).
So, I created test_interface_ixxat_fd.py imitating test_interface_ixxat.py (so that at least the same level of testing was achieved), and in the other hand, to ensure the code I did worked, I used a custom board I had (with personal code) that has canfd connection to talk with ixxat.
With that setup I tried simple messaging works (also with different sets of initialization options), and also more complex cases, like uds messages using udsoncan/can-isotp python libraries, to see that integration is good (I detected a bug in this scenario).
My tests are manual and only work with my board and the embedded code inside, so, I don't think we can use them in this library.
On the other side, for Back2BackTestCase based test, I would need another interface to check that works, but I have only one ixxat.
I'm open to contribute to improve testing for ixxat, but I would need some guidance (to check what's wrong with previous tests), and I would prefer to do it in a different pull request. I'm a bit worried because I saw some of the other pull requests will cause conflicts with this one and it may be hard to solve them...

marcel-kanter replied:
Honestly the tests are a bit buggy and not very well done for the ixxat interface. The current state is an first attempt and felix accidently broke one. So don't worry about that.

felixdivo replied:
Hey @fjburgos that's alright, then we merge this without thorough testing. And sorry, I indeed broke a test which I didn't realize since the CI doesn't run that test. Thanks @marcel-kanter for caching that.

In the end, testing is about finding bugs. If you tested with a manual setup, that's fine (for now) too! If you are interested in more testing, just open an new issue and we can discuss it there. 😄

felixdivo reviewed (changes requested):
A great set of changes!

hardbyte commented:
@fjburgos many thanks for the contribution and for patience with the review process. 

Once the changes requested by @felixdivo are addressed this is good to merge.

fjburgos commented on can/interfaces/ixxat/canlib.py:
```diff
@@ -1,414 +1,40 @@
+        **kwargs
+    ):
         """
         :param int channel:
```
Also changed.

felixdivo replied:
Nice!

fjburgos commented on test/test_interface_ixxat_fd.py:
```diff
@@ -0,0 +1,62 @@
+
+        # rxFifoSize must be > 0
+        with self.assertRaises(ValueError):
+            can.Bus(interface="ixxat", fd=True, channel=0, rxFifoSize=0)
```
Updated.

fjburgos commented on test/test_interface_ixxat_fd.py:
```diff
@@ -0,0 +1,62 @@
+
+        # txFifoSize must be > 0
+        with self.assertRaises(ValueError):
+            can.Bus(interface="ixxat", fd=True, channel=0, txFifoSize=0)
```
Updated.

fjburgos commented:
> @fjburgos Nice work. But the extended argument is not used in the init of vcimpl.py Please set the default to True, because the old implementation opens the interface with
> 
> constants.CAN_OPMODE_EXTENDED
> 
> so extended ids are allowed by default.

I changed the default value to True, and also changed in vcimpl to use the parameter.

fjburgos commented:
> @fjburgos many thanks for the contribution and for patience with the review process.
> 
> Once the changes requested by @felixdivo are addressed this is good to merge.

No problem, I'm just worried about conflicts in case othe pull requests are merged.

I didn't expect this process was so long, but as I said, this is my first contribution to open source project, so, I assume this is normal. In any case, code was improved in each review.

felixdivo commented:
> I didn't expect this process to be so long, but as I said, this is my first contribution to an open source project, so, I assume this is normal. In any case, code was improved in each review.

Well, this is also to a large part caused by us only responding after a few weeks right in the beginning. Thanks for the effort and patience, all!

And in general: Most PRs are indeed a bit quicker, but mostly because they are smaller and more narrow in their scope (my subjective impression). You, however, tackled a rather large change as your first project. Well done :) 
![image](https://user-images.githubusercontent.com/4403130/145090135-346c3cc2-2dfe-439f-938f-6d23d0b1e5c1.png)

cowo78 commented:
I've been working on #1141 that shares the same scope (IXXAT support) and found that canlib_vcinpl2.py duplicates some code from canlib_vcinpl.py. I would suggest some refactoring to create a cleaner codebase.

fjburgos commented:
I would suggest first to improve testing. I did not refactor to avoid risk of introduce bugs on vcinpl side, I only have an fd device, so, I could not check that the previous functionalitites are still working after refactoring.

cowo78 commented:
@felixdivo if #1141 looks fine I propose to create one more PR to remove duplicate code and improve the overall architecture. I can work on such PR and improve testing. @fjburgos would you be available to perform some tests with connected hardware?

felixdivo commented:
Yes, that sounds reasonable. I'll review #1141 soon 👍

fjburgos commented:
I'm available, just say when code is ready and what concrete tests you want me to do.
